src/vsr.zig

```diff
@@ -811,6 +817,228 @@ pub const BlockRequest = extern struct {
+    members: Members,
+    /// The new epoch.
+    ///
+    /// Request is rejected if it isn't exactly current epoch + 1, to protect from operator errors.
+    /// Although there's already an `epoch` field in vsr.Header, we don't want to rely on that for
+    /// reconfiguration itself, as it is updated automatically by the clients, and here we need
+    /// a manual confirmation from the operator.
+    epoch: u32,
+    /// The new replica count.
+    ///
+    /// At the moment, we require this to be equal to the old count.
+    replica_count: u8,
+    /// The new standby count.
+    ///
+    /// At the moment, we require this to be equal to the old count.
+    standby_count: u8,
+    reserved: [10]u8 = [_]u8{0} ** 10,
+    /// The result of this request. Set to zero by the client and filled-in by the primary when it
+    /// accepts a reconfiguration request.
+    result: ReconfigurationResult,
+
+    pub fn validate(
+        request: *const ReconfigurationRequest,
+        current: struct {
+            members: *const Members,
+            epoch: u32,
+            replica_count: u8,
+            standby_count: u8,
+        },
+    ) ReconfigurationResult {
+        assert(valid_members(current.members));
+        assert(member_count(current.members) == current.replica_count + current.standby_count);
+
+        if (request.replica_count == 0) return .replica_count_max_exceeded;
+        if (request.replica_count > constants.replicas_max) return .replica_count_max_exceeded;
+        if (request.standby_count > constants.standbys_max) return .standby_count_max_exceeded;
+
+        if (!valid_members(&request.members)) return .members_invalid;
+        if (member_count(&request.members) != request.replica_count + request.standby_count) {
+            return .members_count_invalid;
+        }
+
+        if (!std.mem.allEqual(u8, &request.reserved, 0)) return .reserved_field;
+        if (request.result != .ok) return .result_must_be_ok;
+
+        if (current.epoch == std.math.maxInt(u32)) return .epoch_overflow;
+        if (request.epoch <= current.epoch) return .epoch_in_the_past;
+        if (request.epoch > current.epoch + 1) return .epoch_in_the_future;
+        assert(request.epoch == current.epoch + 1);
+
+        if (request.replica_count != current.replica_count) return .different_replica_count;
+        if (request.standby_count != current.standby_count) return .different_standby_count;
+
+        // Note: neither set can have duplicates at this point.
+        assert(member_count(current.members) == member_count(&request.members));
+        for (current.members) |member_current| {
+            if (member_current == 0) break;
+            for (request.members) |member| {
+                if (member == member_current) break;
+            } else return .different_member_set;
+        }
+
+        if (std.meta.eql(request.members, current.members.*)) return .no_op;
```

Comment:
Maybe use `std.mem.eql` instead, to avoid copying the arrays onto the stack.
```suggestion
        if (std.mem.eql(u128, &request.members, current.members)) return .no_op;
```